capsules/src/bmp280.rs

```diff
@@ -0,0 +1,378 @@
+    dig_t1: u16,
+    dig_t2: i16,
+    dig_t3: i16,
+    // TODO: pressure calibration
+}
+
+/// CAUTION: calibration data puts least significant byte in the lowest address,
+/// readouts do the opposite.
+fn twobyte(lsb: u8, msb: u8) -> u16 {
+    ((msb as u16) << 8) + lsb as u16
+}
+
+impl CalibrationData {
+    fn new(i2c_raw: &[u8]) -> Self {
+        CalibrationData {
+            dig_t1: twobyte(i2c_raw[0], i2c_raw[1]) as u16,
+            dig_t2: twobyte(i2c_raw[2], i2c_raw[3]) as i16,
+            dig_t3: twobyte(i2c_raw[4], i2c_raw[5]) as i16,
+        }
+    }
+
+    fn temp_from_raw(&self, raw_temp: u32) -> i32 {
+        debug!("raw temp: {}", raw_temp);
+        let temp = raw_temp as i32; // guaranteed to succeed because raw temp has only 20 significant bits maximum.
+        let dig_t1 = self.dig_t1 as i32; // same, 16-bits
+        let dig_t2 = self.dig_t2 as i32; // same, 16-bits
+        let dig_t3 = self.dig_t3 as i32; // same, 16-bits
+                                         // From the datasheet
+        let var1 = (((temp >> 3) - (dig_t1 << 1)) * dig_t2) >> 11;
+        let a = (temp >> 4) - dig_t1;
+        let var2 = (((a * a) >> 12) * dig_t3) >> 14;
+        let t_fine = var1 + var2;
+        ((t_fine * 5) + 128) >> 8
+    }
+}
+
+/// Internal state.
+/// Each state can lead to the next on in order of appearance.
+#[derive(Clone, Copy, PartialEq, Debug)]
+enum State {
+    Uninitialized,
+    InitId,
+    /// It's not guaranteed that the MCU reset is the same as device power-on,
+    /// so an explicit reset is necessary.
+    InitResetting,
+    InitWaitingReady,
+    InitReadingCalibration,
+
+    Idle(CalibrationData),
+
+    // States related to sample readout
+    /// One-shot mode request sent
+    Configuring(CalibrationData),
+    /// Sampling takes milliseconds, so spend most of that time sleeping.
+    WaitingForAlarm(CalibrationData),
+    /// Polling for the readout to become ready.
+    Waiting(CalibrationData),
+    /// Waiting for readout to return.
+    /// This state can also lead back to Idle.
+    Reading(CalibrationData),
+
+    /// The ID is not matching. Reset cannot be attempted.
+    MismatchedDevice,
+    /// Irrecoverable. Currently only when init fails.
+    Error,
+}
+
+/// Complies with the reading and writing protocol used by the sensor.
+struct I2cWrapper<'a> {
+    i2c: &'a dyn i2c::I2CDevice,
+}
+
+impl<'a> I2cWrapper<'a> {
+    fn write<const COUNT: usize>(
+        &self,
+        buffer: &'static mut [u8],
+        addr: Register,
+        data: [u8; COUNT],
+    ) {
+        buffer[0] = addr as u8;
+        for (i, d) in data.iter().enumerate() {
+            buffer[i + 1] = *d;
+        }
+        // TODO: propagate errors
+        self.i2c.enable();
+        self.i2c.write(buffer, COUNT as u8 + 1).unwrap();
+    }
+
+    /// Requests a read into buffer.
+    /// Parse the result using `parse_read`.
+    fn read(&self, buffer: &'static mut [u8], addr: Register, count: u8) {
```

Comment:
Return the buffer in case if an error
```suggestion
    fn read(&self, buffer: &'static mut [u8], addr: Register, count: u8) -> Result<(), (ErrorCode, &'static mut [u8])> {
```